src/main.ts

```diff
@@ -1,256 +1,152 @@
-  }
-
-  map<B>(f: (a: A) => B): Dataway<L, B> {
-    return this as any;
-  }
-  ap<B>(fab: Dataway<L, (a: A) => B>): Dataway<L, B> {
-    if (fab.isFailure()) {
-      return fab as any;
-    }
-    return (fab.isNotAsked() ? fab : this) as any;
-  }
-
-  chain<B>(f: (a: A) => Dataway<L, B>): Dataway<L, B> {
-    return this as any;
-  }
-
-  fold<B>(
-    onNotAsked: B,
-    onLoading: B,
-    onFailure: (e: L) => B,
-    onSuccess: (a: A) => B,
-  ) {
-    return onNotAsked;
-  }
+export interface NotAsked {
+  readonly _tag: 'NotAsked';
 }
 
-export class Loading<L, A> {
-  readonly _tag = 'Loading';
-  constructor() {}
-
-  isNotAsked() {
-    return false;
-  }
-  isLoading() {
-    return true;
-  }
-  isFailure() {
-    return false;
-  }
-  isSuccess() {
-    return false;
-  }
-
-  map<B>(f: (a: A) => B): Dataway<L, B> {
-    return this as any;
-  }
-  ap<B>(fab: Dataway<L, (a: A) => B>): Dataway<L, B> {
-    if (fab.isNotAsked()) {
-      return notAsked();
-    }
-    if (fab.isFailure()) {
-      return fab as any;
-    }
-    return (fab.isLoading() ? fab : this) as any;
-  }
-
-  chain<B>(f: (a: A) => Dataway<L, B>): Dataway<L, B> {
-    return this as any;
-  }
-
-  fold<B>(
-    onNotAsked: B,
-    onLoading: B,
-    onFailure: (e: L) => B,
-    onSuccess: (a: A) => B,
-  ) {
-    return onLoading;
-  }
+export interface Loading {
+  readonly _tag: 'Loading';
 }
 
-export class Failure<L, A> {
-  readonly _tag = 'Failure';
-  constructor(readonly value: L) {}
-
-  isNotAsked() {
-    return false;
-  }
-  isLoading() {
-    return false;
-  }
-  isFailure() {
-    return true;
-  }
-  isSuccess() {
-    return false;
-  }
-
-  map<B>(f: (a: A) => B): Dataway<L, B> {
-    return this as any;
-  }
-  ap<B>(fab: Dataway<L, (a: A) => B>): Dataway<L, B> {
-    return (fab.isFailure() ? fab : this) as any;
-  }
-
-  chain<B>(f: (a: A) => Dataway<L, B>): Dataway<L, B> {
-    return this as any;
-  }
+export interface Failure<F> {
+  readonly _tag: 'Failure';
+  readonly failure: F;
+}
 
-  fold<B>(
-    onNotAsked: B,
-    onLoading: B,
-    onFailure: (e: L) => B,
-    onSuccess: (a: A) => B,
-  ) {
-    return onFailure(this.value);
-  }
+export interface Success<S> {
+  readonly _tag: 'Success';
+  readonly success: S;
 }
 
-export class Success<L, A> {
-  readonly _tag = 'Success';
-  constructor(readonly value: A) {}
+export type Dataway<F, S> = NotAsked | Loading | Failure<F> | Success<S>;
 
-  isNotAsked() {
-    return false;
-  }
-  isLoading() {
-    return false;
-  }
-  isFailure() {
-    return false;
-  }
-  isSuccess() {
-    return true;
-  }
-
-  map<B>(f: (a: A) => B): Dataway<L, B> {
-    return new Success(f(this.value));
-  }
-  ap<B>(fab: Dataway<L, (a: A) => B>): Dataway<L, B> {
-    if (fab.isSuccess()) {
-      const _fab = fab as Success<L, (a: A) => B>;
-      return this.map(_fab.value);
-    }
-    return fab as any;
-  }
+export const notAsked: Dataway<never, never> = { _tag: 'NotAsked' };
```

Comment:
```suggestion
export const notAsked: Dataway<never, never> = { _tag: 'NotAsked' };
```

T'es certain pour les never là ? c'est pas plutôt `any` que tu voulais mettre.

Je pense qu'avec `never` on pourra jamais mettre un `notAsked` dans un `Dataway<Error, Entity>` par exemple mais je peux mal comprendre

Reply:
Ou bien on peut mettre un unknown remarque non ?